### PR TITLE
Dunification

### PR DESCRIPTION
**Pull request by LaurentMazare**
This PR provides some dune files so that the main library can be build via dune, the tests can also be run with `dune runtest`. The existing makefile system still works.

This also involved cleaning some warnings for things to compile, including a small tweak to the generator to avoid generating multiple functions with the same name.

A non-ideal point is that the main pyml library is marked `(wrapped false)` so that it does not need to be opened by tests. This is probably not much of an issue but should be fixed if we don't need the makefile support for running tests at some point. Also it would be a bit cleaner to move the modules to different subdirectories (which would avoid the `modules` bits in the dune file) but this is probably not straightforward to have while keeping makefile compatibility.

### CODE REVIEW

**Comment by LaurentMazare** [comment]
@andrewray do you want to push with your separate dunification feature or are you happy with this one ?
From what I understand, your version is a more significative change (e.g. also moves the code to subdirectories and tweaks the tests), so I don't have a strong opinion on this.

File: numpy.ml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-external get_pyarray_type: Py.Object.t -> Py.Object.t = "get_pyarray_type"
+external _get_pyarray_type: Py.Object.t -> Py.Object.t = "get_pyarray_type"
```
Comment (thierry-martinez) [inline numpy.ml]: This definition can be removed completely: it is already bound to `Py.Array.pyarray_type`.

File: py.ml [hunk 2]
```diff
@@ -1014,10 +1016,10 @@ let string_length = String.length
 module String__ = struct
   include String_
 
-  let check_bytes s =
+  let _check_bytes s =
```
Comment (thierry-martinez) [inline py.ml]: You may consider to add to `py.mli`:
```diff
+  val check_bytes: Object.t -> bool
+  (** [check_bytes o] returns [o] if [o] is a Python bytes string. *)
```

File: py.ml [hunk 3]
```diff
@@ -1014,10 +1016,10 @@ let string_length = String.length
...
     Type.get s = Type.Bytes
 
-  let check_unicode s =
+  let _check_unicode s =
```
Comment (thierry-martinez) [inline py.ml]: Can be defined in `py.mli` as well:
```diff
+  val check_unicode: Object.t -> bool
+  (** [check_unicode o] returns [o] if [o] is a Python unicode string. *)
```

File: py.ml [hunk 4]
```diff
@@ -1494,7 +1496,7 @@ module Number = struct
 
   let number_and v0 v1 = check_not_null (Pywrappers.pynumber_and v0 v1)
 
-  let check v = Pywrappers.pynumber_check v <> 0
+  let _check v = Pywrappers.pynumber_check v <> 0
```
Comment (thierry-martinez) [inline py.ml]: Why hiding Py.Number.check from the interface (I see you removed it from `py.mli` as well)?

**Comment by thierry-martinez** [comment]
Thank you very much for working on this!

File: py.ml (hunk 4 above)
Comment (LaurentMazare) [reply py.ml]: Well actually there were two `check` function in this module and two in the mli too so only the later one which I haven't removed was used.

**Comment by LaurentMazare** [comment]
@thierry-martinez thanks for the feedback, I pushed some fixes for all but one (the removed `check` function was removed as there were two such functions in the same module so only the second one was ever used).

**Comment by thierry-martinez** [comment]
Thank you very much, @LaurentMazare ! I am ready to merge right now. Should I wait for @andrewray's answer or shall I merge now?

**Comment by andrewray** [comment]
It's probably easier to merge Laurents version as it's a bit simpler - I re-organised the directory structure which is nicer, but makes maintaining the old build much worse - perhaps we can reconsider when the old build gets deprecated.

**Pull request merged**